### ISSUE TITLE
[Discussion] Shared libs dependencies

### ISSUE THREAD

**Issue by es20490446e** [chaotic-aur/packages#2196]
### What happens?

Okay guys. I'm having a problem and perhaps some of you know the answer, and we can even help each other.

I decided to post it here instead of Telegram because this could get a little bit long.

The problem I'm facing is this: I have a personal repository with a few packages. The problem comes when some of those packages depend on shared libraries present in the official Arch Linux repositories.

For preventing sonames breakages in those I decided to package them inside my own repository, but by including older libraries too in the same package. Hence keeping certain compatibility with packages built at different times and repositories.

But then I get an additional problem: what happens if the same package that provides the libs updates sooner from Arch, for example if the PKGBUILD in my repository stops working for a little bit.

Those libs will upgrade from Arch, and all the dependent packages will be broken.

I could use a rather high epoch in those shared libs packages, but that will then break Arch applications if my repo package updates sooner. So even worse.

Or I could make packages depend on specific sonames. But if some break for a while it can hold back entire system upgrades, preventing also from installing individual applications. So a no no.

Basically I haven't figured out yet how to fix this in a no so messy manner. If you have any suggestion it may help me a lot. Thanks in advance.

@dr460nf1r3 @PedroHLC @lordkitsuna @SolarAquarion @thotypous @RustemB @namanlp @IslandC0der @JustTNE @Edu4rdSHL @alexjp @technetium1

### What is expected to happen?

Applications to always find the shared libraries they depend on.

**Comment by es20490446e** [comment]
I'm closing just this not appears as an issue of Chaotic.

**Comment by FabioLolix** [comment]
> The problem I'm facing is this: I have a personal repository with a few packages. The problem comes when some of those packages depend on shared libraries present in the official Arch Linux repositories.

Rebuild that packages, that's how is done in the Arch repo

**Comment by PedroHLC** [comment]
> Those libs will upgrade from Arch, and all the dependent packages will be broken.

In the past, we did try to hash all our dependencies versions and suffix this into the pkgrel (like here https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=bluespec-git#n16) but of course, that was far from a "ArchLinux-native" experience.

I personally solved this, by moving to NixOS, where every dependency's PKGBUILD is nested-hashed into the final built package. :sweat_smile:

**Comment by es20490446e** [comment]
> Rebuild that packages, that's how is done in the Arch repo.

But that doesn't prevent the packages from being broken meanwhile, does it?

> Every dependency's PKGBUILD is nested-hashed into the final built package.

Doesn't that defeat the purpose of having shared libraries in the first place?

I have seen packages grow ten times when doing things like that, plus you now cannot upgrade fixes in a single place.

You are forced to rebuild the package for every dependency upgrade, even when there are no soname changes whatsoever. Plus you have to track such upgrades for every single dependency, not only the software's version.

**Comment by alexjp** [comment]
> But that doesn't prevent the packages from being broken meanwhile, does it?

Only while you don't rebuild them.

> Doesn't that defeat the purpose of having shared libraries in the first place?
>
>I have seen packages grow ten times when doing things like that, plus you now cannot upgrade fixes in a single place.

He is talking about something different, you are thinking of building static binaries. Look up Nixos and Guix. In Nixos/Guix it doesn't necessarily grow 10 times because it is still shared linking, and they are, in most cases shared with the rest of the system.


The situation you have, is exactly why people use Gentoo (revdep rebuild, auto checking of this issue and slots), Nixos (store/hashes), Guix (store/hashes), Flatpak(_lets download a fedora iso for each package, why not go full vm's for each file??_), or static binaries.

You aren't doing anything wrong, is just the way it is.

( disclaimer: the text in italic is an obvious exaggeration of mine )

**Comment by es20490446e** [comment]
Maybe I can just update daily the repo, and that would be good enough for preventing conflicts.

As those shared library packages usually stay in testing for a few weeks.
